frontend/src/app/shared/components/storages/file-picker-base-modal.component.ts/file-picker-base-modal.component.ts

```diff
@@ -0,0 +1,153 @@
+import { SortFilesPipe } from 'core-app/shared/components/storages/pipes/sort-files.pipe';
+import { StorageFilesResourceService } from 'core-app/core/state/storage-files/storage-files.service';
+import { Breadcrumb, BreadcrumbsContent } from 'core-app/spot/components/breadcrumbs/breadcrumbs-content';
+import {
+  StorageFileListItem,
+} from 'core-app/shared/components/storages/storage-file-list-item/storage-file-list-item';
+import {
+  isDirectory,
+  getIconForStorageType,
+  makeFilesCollectionLink,
+} from 'core-app/shared/components/storages/functions/storages.functions';
+
+@Directive()
+export abstract class FilePickerBaseModalComponent extends OpModalComponent implements OnInit, OnDestroy {
+  public breadcrumbs:BreadcrumbsContent;
+
+  public listItems$:Observable<StorageFileListItem[]>;
+
+  public readonly loading$ = new BehaviorSubject<boolean>(true);
+
+  protected get storageLink():IHalResourceLink {
+    return this.locals.storageLink as IHalResourceLink;
+  }
+
+  protected readonly storageFiles$ = new BehaviorSubject<IStorageFile[]>([]);
+
+  private loadingSubscription:Subscription;
+
+  protected constructor(
+    @Inject(OpModalLocalsToken) public locals:OpModalLocalsMap,
+    readonly elementRef:ElementRef,
+    readonly cdRef:ChangeDetectorRef,
+    protected readonly sortFilesPipe:SortFilesPipe,
+    protected readonly storageFilesResourceService:StorageFilesResourceService,
+  ) {
+    super(locals, cdRef, elementRef);
+  }
+
+  ngOnInit():void {
+    super.ngOnInit();
+
+    this.breadcrumbs = new BreadcrumbsContent([{
+      text: this.locals.storageName as string,
+      icon: getIconForStorageType(this.locals.storageType as string),
+      navigate: () => this.changeLevel(null, this.breadcrumbs.crumbs.slice(0, 1)),
+    }]);
+
+    this.listItems$ = this.storageFiles$
+      .pipe(
+        map((files) =>
+          this.sortFilesPipe.transform(files)
+            .map((file, index) => this.storageFileToListItem(file, index))),
+      );
+
+    this.storageFilesResourceService.files(makeFilesCollectionLink(this.storageLink, null))
+      .pipe(take(1))
+      .subscribe((files) => {
+        this.storageFiles$.next(files);
+        this.loading$.next(false);
+      });
+  }
+
+  ngOnDestroy():void {
+    super.ngOnDestroy();
+
+    this.storageFilesResourceService.reset();
+  }
+
+  public openStorageLocation():void {
+    window.open(this.locals.storageLocation, '_blank');
+  }
+
+  protected abstract storageFileToListItem(file:IStorageFile, index:number):StorageFileListItem;
+
+  protected enterDirectoryCallback(directory:IStorageFile):() => void {
+    if (!isDirectory(directory.mimeType)) {
+      return () => {};
+    }
+
+    return () => {
+      const crumbs = this.breadcrumbs.crumbs;
+      const end = crumbs.length + 1;
+      const newCrumb:Breadcrumb = {
+        text: directory.name,
+        navigate: () => this.changeLevel(directory.location, this.breadcrumbs.crumbs.slice(0, end)),
```

Comment:
```suggestion
        navigate: () => this.changeLevel(directory.location, crumbs.slice(0, end)),
```

Reply:
this suggested change actually is a bug, I made it myself when implementing it first. Problem here is the closure. the `navigate` callback MUST use the `this` instance (of the future state of the modal). Because in `crumbs` object, the `newCrumb` is not yet included, but it must be the target of the navigation (hence `end = crumbs.length + 1`)

Reply:
Oh I see. Perhaps we should add a comment to make that more clear

Reply:
I'll add a comment